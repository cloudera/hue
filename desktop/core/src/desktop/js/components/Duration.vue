<!--
  Licensed to Cloudera, Inc. under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  Cloudera, Inc. licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<template>
  <span>{{ duration }}</span>
</template>

<script lang="ts">
  import { defineComponent } from 'vue';

  import { Duration as LuxonDuration, DurationUnit } from 'luxon';
  import I18n from '../utils/i18n';

  const SHORT_LIMITS = [
    { unit: <DurationUnit>'years', postfix: 'year', appendS: true },
    { unit: <DurationUnit>'days', postfix: 'day', appendS: true },
    { unit: <DurationUnit>'hours', postfix: 'h', appendS: false },
    { unit: <DurationUnit>'minutes', postfix: 'm', appendS: false },
    { unit: <DurationUnit>'seconds', postfix: 's', appendS: false },
    { unit: <DurationUnit>'milliseconds', postfix: 'ms', appendS: false }
  ];

  export const duration = (value: number, short?: boolean): string => {
    const luxonDuration = LuxonDuration.fromMillis(value);
    if (short) {
      for (const limit of SHORT_LIMITS) {
        const val = Math.floor(luxonDuration.as(limit.unit));
        if (val >= 1) {
          return val + I18n(limit.postfix + (limit.appendS && val > 1 ? 's' : ''));
        }
      }
      return `0${I18n('ms')}`;
    }
    return luxonDuration.toFormat('hh:mm:ss');
  };

  export default defineComponent({
    name: 'Duration',
    props: {
      value: {
        type: Number,
        required: true
      },
      short: Boolean
    },

    computed: {
      duration(): string {
        return duration(this.value, this.short);
      }
    }
  });
</script>

<style lang="scss" scoped></style>
