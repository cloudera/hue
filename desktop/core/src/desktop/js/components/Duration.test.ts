// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// 'License'); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallowMount } from '@vue/test-utils';
import Duration, { duration } from './Duration.vue';

describe('Duration.vue', () => {
  it('should render', () => {
    const wrapper = shallowMount(Duration, {
      propsData: {
        value: 1234567890
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render short format', () => {
    const wrapper = shallowMount(Duration, {
      propsData: {
        value: 1234567890,
        short: true
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render 00:00:00 duration', () => {
    const wrapper = shallowMount(Duration, {
      propsData: {
        value: 0
      }
    });
    expect(wrapper.element.firstChild?.textContent).toContain('00:00:00');
  });

  it('should render 01:01:01 duration', () => {
    const wrapper = shallowMount(Duration, {
      propsData: {
        value: 1000 + 1000 * 60 + 1000 * 60 * 60
      }
    });
    expect(wrapper.element.firstChild?.textContent).toContain('01:01:01');
  });

  it('should render 100:01:01 duration', () => {
    const wrapper = shallowMount(Duration, {
      propsData: {
        value: 1000 + 1000 * 60 + 1000 * 60 * 60 * 100
      }
    });
    expect(wrapper.element.firstChild?.textContent).toContain('100:01:01');
  });

  it('should render 4days short duration', () => {
    expect(duration(1000 * 60 * 60 * 24 * 4, true)).toEqual('4days');
  });
});
