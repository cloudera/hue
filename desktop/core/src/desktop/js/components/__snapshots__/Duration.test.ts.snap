// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Duration.vue should render 1`] = `
<span>
  342:56:07
</span>
`;

exports[`Duration.vue should render short format 1`] = `
<span>
  14days
</span>
`;
